pkg/utils/taints.go

```diff
@@ -0,0 +1,110 @@
+package utils
+
+// Copy paste from SNR - https://github.com/medik8s/self-node-remediation/blob/main/pkg/utils/taints.go
+import (
+	"context"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/medik8s/fence-agents-remediation/api/v1alpha1"
+)
+
+var (
+	loggerTaint = ctrl.Log.WithName("taints")
+)
+
+// taintExists checks if the given taint exists in list of taints. Returns true if exists false otherwise.
+func taintExists(taints []corev1.Taint, taintToFind *corev1.Taint) bool {
+	for _, taint := range taints {
+		if taint.MatchTaint(taintToFind) {
+			return true
+		}
+	}
+	return false
+}
+
+// deleteTaint removes all the taints that have the same key and effect to given taintToDelete.
+func deleteTaint(taints []corev1.Taint, taintToDelete *corev1.Taint) ([]corev1.Taint, bool) {
+	var newTaints []corev1.Taint
+	deleted := false
+	for i := range taints {
+		if taintToDelete.MatchTaint(&taints[i]) {
+			deleted = true
+			continue
+		}
+		newTaints = append(newTaints, taints[i])
+	}
+	return newTaints, deleted
+}
+
+// createNoExecuteTaint returns a remediation NoExeucte taint
+func createNoExecuteTaint() corev1.Taint {
+	return corev1.Taint{
+		Key:    v1alpha1.Medik8sRemediationTaintKey,
+		Value:  v1alpha1.FARRemediationTaintValue,
+		Effect: corev1.TaintEffectNoExecute,
+	}
+}
+
+// AppendTaint appends new taint to the taint list when it is not present, and returns error if it fails in the process
+func AppendTaint(r client.Client, nodeName string) error {
+	// find node by name
+	node, err := getNodeWithName(r, nodeName)
+	if err != nil {
+		return err
+	}
+
+	taint := createNoExecuteTaint()
+	// check if taint doesn't exist
+	if taintExists(node.Spec.Taints, &taint) {
+		return nil
+	}
+	// add the taint to the taint list
+	patch := client.MergeFrom(node.DeepCopy())
```

Comment:
Using `patch` for updating taints is very risky unfortunately. The API server is updating taints as well while nodes get healthy/unhealthy. We saw this with SNR IIRC, when it was overwriting changes made by the API server, or the other way around. So please either use `Update` below, or `MergeFromWithOptions` here with the `OptimisticLock` option. Same for remove taint.